lib/frontend/src/Obelisk/Frontend/Cookie.hs

```diff
@@ -50,6 +96,119 @@ instance HasCookies m => HasCookies (RoutedT t r m)
+      B64.decode c
+
+data GetCookieFailed
+  = GetCookieFailed_NotFound
+  | GetCookieFailed_Base64DecodeFailed String
+  deriving (Eq, Show, Read)
+
+-- | Contexts in which cookies can be set. It's best to use the Base64
+-- encoding 'setCookie' function over 'setCookieRaw', because the
+-- former doesn't fail silently on unsupported characters. Be warned
+-- that browsers still limit the (total) size of cookies. This limit
+-- is unchecked.
+--
+-- As of writing this class is implemented in the following contexts:
+--
+-- * Backend, on the 'Snap.Core.Snap' monad where it sets cookies for the HTTP response.
+--
+-- * Frontend, on 'Reflex.Dom.Core.Client' (this is the "Client m a"
+--   part of 'Reflex.Dom.Prerender.prerender'). This sets cookies
+--   client-side.
+class (Monad m) => HasSetCookie m where
+  setCookieRaw :: SetCookie -> m ()
+  default setCookieRaw :: (HasSetCookie m', m ~ t m', MonadTrans t) => SetCookie -> m ()
+  setCookieRaw c = lift (setCookieRaw c)
+
+instance (MonadJSM m, RawDocument (DomBuilderSpace (HydrationDomBuilderT s t m)) ~ Document) => HasSetCookie (HydrationDomBuilderT s t m) where
+  setCookieRaw cookie = do
+    doc <- askDocument
+    liftJSM $ DOM.setCookie doc $ decodeUtf8 $ LBS.toStrict $ toLazyByteString $ renderSetCookie cookie
+
+
+instance (MonadJSM m, HasDocument m, DOM.IsDocument (RawDocument (DomBuilderSpace m))) => HasSetCookie (HydratableT m) where
+  setCookieRaw cookie = do
+    doc <- askDocument
+    liftJSM $ DOM.setCookie doc $ decodeUtf8 $ LBS.toStrict $ toLazyByteString $ renderSetCookie cookie
+
+
+instance HasSetCookie m => HasSetCookie (BehaviorWriterT t w m)
+instance HasSetCookie m => HasSetCookie (DynamicWriterT t w m)
+instance HasSetCookie m => HasSetCookie (EventWriterT t w m)
+instance HasSetCookie m => HasSetCookie (PostBuildT t m)
+instance HasSetCookie m => HasSetCookie (QueryT t q m)
+instance HasSetCookie m => HasSetCookie (ReaderT r m)
+instance HasSetCookie m => HasSetCookie (RequesterT t request response m)
+instance HasSetCookie m => HasSetCookie (RouteToUrlT t m)
+instance HasSetCookie m => HasSetCookie (SetRouteT t r m)
+instance HasSetCookie m => HasSetCookie (StaticDomBuilderT t m)
+instance HasSetCookie m => HasSetCookie (TriggerEventT t m)
+instance HasSetCookie m => HasSetCookie (RoutedT t r m)
+instance HasSetCookie m => HasSetCookie (ConfigsT m)
+instance HasSetCookie m => HasSetCookie (CookiesT m)
+instance (MonadJSM m, RawDocument (DomBuilderSpace (HydrationDomBuilderT s t m)) ~ Document) => HasCookies (HydrationDomBuilderT s t m) where
+  askCookies = fmap (parseCookies . encodeUtf8) $ DOM.getCookie =<< askDocument
+
+instance (MonadJSM m, HasDocument m, DOM.IsDocument (RawDocument (DomBuilderSpace m))) => HasCookies (HydratableT m) where
+  askCookies = fmap (parseCookies . encodeUtf8) $ DOM.getCookie =<< askDocument
+
+-- | Store a cookie which will be Base64 encoded.
+setCookie :: (HasSetCookie m) => SetCookie -> m ()
```

Comment:
Can we name this `encodeAndSetCookie`? I don't want it to be surprising that this function re-encodes your cookie value for you.